### ISSUE TITLE
Tracer Bullet

### ISSUE THREAD

**Issue by SimonBiggs** [libreapp/libreapp#6]
The initial tracer bullet will have the web app automatically print a DICOM header when a DICOM file is placed in a defined directory on the local machine.

This will be achieved by setting up a local server, the local server and the web app will open up a websocket communication. The local server will run the watchdog library. When the DICOM file is dropped in the defined directory, this will trigger watchdog, send the DICOM file to the web app, the web app will then trigger a DICOM load and print via pyodide.

No user interaction within the webapp needed.

**Comment by darcymason** [comment]
Nice!  Sounds like a good complete initial setup.

About next stages ... any thoughts on pipelines?  I've been thinking about how we could set up a pipeline of multiple processing steps through separate tools.  It could be by files (probably necessary in some cases) but I wonder also about some kind of pipeline communication between stages - json dicts, perhaps? If files needed, the filenames could be identified also in the json.  The pipeline idea could be an incremental step forward from this first tracer.

**Comment by SimonBiggs** [comment]
I haven't given it thought no. And a picture isn't immediately forming in my head for how best to do it....

**Comment by SimonBiggs** [comment]
For files we might want users to be editing I'm keen to use the toml format:

https://github.com/toml-lang/toml

**Comment by darcymason** [comment]
toml looks reasonable.  Just like .ini if it is simple enough, which people are often used to or can grok very quickly.  But if I understand where things are going, I would hope that editing files shouldn't be necessary.  If the local server can be set up, then I'm thinking a web-based dashboard could be used for config.  I guess we need to see the tracer in place to start to make all this real...

And the pipeline concept -- I was thinking much like linux tools just pipe from one into the other, users (or us) could connect tools together according to the needs.  For example, a Winston-Lutz tool analyzes a set of files and produces coordinates (and a date-time, and a unit id etc.), which can then plug into a comparison with something else, or a database for storage.  I don't have a complete picture in my head yet either, but just throwing the ideas out there.

**Comment by SimonBiggs** [comment]
Yup, I do really like that concept.

**Comment by Matthew-Jennings** [comment]
I like the pipeline concept! Also agree that files will be needed in some cases. I suspect certain stages would benefit from supporting both.

Putting the legal issues aside for the moment, an excellent example is DICOM anonymisation. It would be a very common early stage for any DICOM related pipeline, but is also desirable as a standalone process. Is there any issue with supporting both TOML (say) and file inputs/outputs?

**Comment by Matthew-Jennings** [comment]
Maybe I'm missing something, but what is the need for pyodide if there is a local server running a (presumably python based) watchdog? I.e., does that local server have Python installed? Or is the watchdog being executed via pyodide?

Moreover, if there's no user interaction with the "web app", why not just connect to a CLI and have the DICOM header printed to, say, a text file?

**Comment by SimonBiggs** [comment]
It's the no installation point. The aim is to have all applications work by default with nothing installed on the user's computer. The user can then choose to install something if they want to have extra benefits (such as automation). But even when a server is installed it will run through the same application code, just be triggered by os events instead of mouse clicks...

**Comment by Matthew-Jennings** [comment]
> It's the no installation point.

But if I’m reading this correctly, the user needs to have Python installed on a local server to run the watchdog. So... we’re not circumventing installation? And if Python exists on a local server, there’s no need for client-side execution (pyodide)...

What am I missing?

**Comment by Matthew-Jennings** [comment]
Wait, the watchdog is the optional extra? We’re just making sure the pyodide-based web-app can cater to that?

That makes sense, sorry.

**Comment by SimonBiggs** [comment]
Yup, watchdog is the optional extra. And it'll likely be distributed in a "frozen" format. Separate from any locally installed python if there is one.

No need to be sorry :)

**Comment by SimonBiggs** [comment]
Also, if we follow the route I'm currently following it won't be on a local server anymore. Instead it will be on the local machine. Each desktop computer that wants to have the automation functionality will need to have the browser extension and the local automation application installed.

**Comment by SimonBiggs** [comment]
> I like the pipeline concept! Also agree that files will be needed in some cases. I suspect certain stages would benefit from supporting both.
> 
> Putting the legal issues aside for the moment, an excellent example is DICOM anonymisation. It would be a very common early stage for any DICOM related pipeline, but is also desirable as a standalone process. Is there any issue with supporting both TOML (say) and file inputs/outputs?

I think having an automated DICOM anonymiser would be a great next step after the initial "print a header".

To get to that point I'll want to begin iterating on the platform syntax itself. As in, what bare minimum files are required to make an application. 

* A set of react files for the user interface
* Some python files for the logic
* An optional logo?
* And maybe a top level configuration file with the following
  * Pipeline contract (input/output)
  * Name, description
  * What features it implements
  * Any other configuration options we might need in the future
  * What categories it fits under

Will need to design a filterable navigation panel, as one can imagine a user having 100s of these applications. Need to be able to click tags as a way to filter by category, type to filter by text.

**Comment by darcymason** [comment]
> Yup, watchdog is the optional extra. And it'll likely be distributed in a "frozen" format. Separate from any locally installed python if there is one.

Just wanted to add that even if the local watchdog does become necessary, it should be the absolute minimum needed to get around the restricted browser rights issues - e.g. ability to read/write local files or interact with local db server. Then maintenance/updates etc are so much easier because there is very little local install (hopefully 'frozen' forever as Simon noted).

> Also, if we follow the route I'm currently following it won't be on a local server anymore. Instead it will be on the local machine. Each desktop computer that wants to have the automation functionality will need to have the browser extension and the local automation application installed

That looks very promising. How is security handled, though?  How does the app only allow our requests and not other, malicious web pages?

**Comment by SimonBiggs** [comment]
> That looks very promising. How is security handled, though? How does the app only allow our requests and not other, malicious web pages?

There is an "allowed_origins" tag which declares which extension(s) can access the app:

https://developer.chrome.com/apps/nativeMessaging#native-messaging-host

That extension will then only interact with the defined web pages. (I believe...)

**Comment by SimonBiggs** [comment]
Initial tracer bullet working in https://github.com/libreapp/libreapp/pull/5

I decided to go with non-encrypted websockets talking to 127.0.0.1. This is a trade off.

Here were my options:

Use NativeMessaging:

* Pros
  * Works with all browsers without needing to install certificates
* Cons
  * Installation method for each browser and OS is different
    * results in 9 installation methods if we want to support 3 OSs and 3 browsers
    * This seems unreasonable to maintain
  * Limited to only sending 1 Mb at a time
  * Have to convert binary to string to send to browser, then convert back to string
  * Documentation is lacking
  * Quite a new protocol

Use XHR

* Pros
  * Works with all browsers without needing to install certificates
* Cons
  * Have to convert binary to string to send to browser, then convert back to string again
  * Doesn't support two way communication simply

Use websockets

* Pros
  * Allows binary transfer
  * Allows asynchronous two way transfer between server and browser
  * Well documented and supported
* Cons
  * Only browsers that are chromium based allow access to websockets on IP 127.0.0.1 without requiring a certificate
    * Of the main browsers in circulation this includes Edge and Chrome
    * Specifically, Firefox does not support this without installing a certificate
    * Certificates are not insurmountable to install, just a little difficult.

In the end, I chose to go with websockets.
